src/compute/src/rpc/service/stream_service.rs

```diff
@@ -124,11 +124,12 @@ impl StreamService for StreamServiceImpl {
         request: Request<ForceStopActorsRequest>,
     ) -> std::result::Result<Response<ForceStopActorsResponse>, Status> {
         let req = request.into_inner();
-        self.mgr.stop_all_actors().await?;
+        let previous_actor_failure_cause = self.mgr.stop_all_actors().await;
```

Comment:
> Meta node simply logs these errors received from each compute node. This can be further refined to identify and only log the wanted one.

Agree with this. So why not sorting the errors on Meta node? In that way, the compute nodes simply report all errors to meta, and the meta node can pick the "root" one.

Reply:
Doable. But in case there are lots of actors, and we want to keep backtrace, then the total data size to meta node can be large.

Reply:
Agree, we should do some local sort in CN first, then only return a single error to meta.

Reply:
Do we really need the sorting? 😕 Note that local sorting means buffering. We're not sure how many errors we should process or how long we should wait, which inevitably leads to returning error in `ForceStopActorsResponse`.

A simple proposal is to use gRPC streaming to return errors, filtering out some of them to ensure that the ranking of these errors **strictly** increases. When the meta node finds an error interesting or when the timeout has been reached, we can simply cut off the `CollectBarrier` RPC connection and we're done.

Reply:
I still recommend maintaining and distinguishing actor errors on the compute nodes, and returning them through the `CollectBarrier` RPC. On the CN, we can return errors at the following time points: 1. After receiving all reported errors from actors; 2. After receiving the first error and timing out. This way, corresponding streaming DDL with `Foreground` can also directly display errors on the client side.
To avoid buffering the errors, we can simply count the `Channel closed` error instead. Or even we can group and count according to different types of errors, and only buffer one copy of the error, if our streaming error can be well classified.

Reply:
With regards to handle errors of multiple compute nodes in meta node, as [ServerError](https://github.com/risingwavelabs/risingwave/blob/80045051a7de3e0a7249b26a33ace96618f5c95b/src/error/src/tonic.rs#L54) has erased inner error type, it's harder to implement a `try_find_root_cause` in meta node than in compute node. Although currently the [service_name](https://github.com/risingwavelabs/risingwave/blob/998b1c90a52b3c0e44ea32a9a9510a75bb197198/src/stream/src/error.rs#L179) field has an 1-1 mapping to each concrete error type, I suppose we shouldn't rely on it.

That's why I just keep all compute nodes' errors instead. I think it's still OK to check several compute node's error, as long as each is concise.

Reply:
> I still recommend maintaining and distinguishing actor errors on the compute nodes, and returning them through the CollectBarrier, SendBarrier RPCs. On the CN, we can return errors at the following time points: 1. After receiving all reported errors from actors; 2. After receiving the first error and timing out.

This also LGTM and be much easier to implement. Just add another timeout in current implementation.

> To avoid buffering the errors

Regarding "buffering the errors in compute node", as we already do it so it's not a big deal?